### PR TITLE
http3: validating codec

### PR DESCRIPTION
**Pull request by alyssawilk**
Rejecting invalid config which only partially configures HTTP/3

Risk Level: low (validates broken config)
Testing: new unit tests
Docs Changes: n/a
Release Notes: n/a
Fixes https://github.com/envoyproxy/envoy/issues/15985

### CODE REVIEW

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -930,6 +930,16 @@ Network::DrainableFilterChainSharedPtr ListenerFilterChainFactoryBuilder::buildF
                                      "{}. \nUse QuicDownstreamTransport instead.",
                                      transport_socket.DebugString()));
   }
+  const std::string config_str =
+      filter_chain.filters_size() == 0 ? "" : filter_chain.filters(0).DebugString();
+  const std::string hcm_str =
+      "envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager";
+  if (is_quic && (filter_chain.filters().size() != 1 || !absl::StrContains(config_str, hcm_str) ||
+                  !absl::StrContains(config_str, "codec_type: HTTP3"))) {
```
Comment (danzh2010) [inline source/server/listener_manager_impl.cc:938]: StrContains() is inefficient and this doesn't prevent other codec types from being misconfigured to HTTP3.
I have some thoughts about validating hcm and codec_type here:
We can compare filter_chain.filters(0).typed_config().type_url() here with "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager" to validate HCM is the only network filter configured.

As to codec_type, the caller of HttpConnectionManagerConfig::createCodec() -- Http::ConnectionManagerImpl-- actually knows if the connection is quic or not. Can we modify createCodec() interface to take in a bool and always create a Http3 codec if it's true? If the pass-in value isn't consistent with configured codec type, i.e. createCodec(/*is_quic*/true) with codec_type_ != HTTP3 or createCodec(/*is_quic*/false) with codec_type_ == HTTP3, create whatever makes the most sense and warn about that.

Comment (alyssawilk) [reply source/server/listener_manager_impl.cc:938]: switching to hcm type url makes sense. 
I'm not worried about an expensive string compare for config load - it's infrequent and already expensive.  I think ignoring configured codec type would not be Ok - it'd better to reject invalid config than accept, do something unexpeced and runtime-warn about it.

Comment (mattklein123) [reply source/server/listener_manager_impl.cc:940]: I don't love doing the unpack twice, but why not unpack the HCM config into a real object, and then do the check in code? That would be a lot less hacky than doing str contains on a debug string?

Another option if we don't want to unpack twice. What about doing this check in the HCM config load itself? We could probably figure out a way to get back to the listener config and do the check there?

/wait-any

**Review by mattklein123 — APPROVED** [review]
Perfect, thanks!

File: source/server/listener_manager_impl.cc [hunk 2]
```diff
@@ -930,18 +930,6 @@ Network::DrainableFilterChainSharedPtr ListenerFilterChainFactoryBuilder::buildF
                                      "{}. \nUse QuicDownstreamTransport instead.",
                                      transport_socket.DebugString()));
   }
-  const std::string config_str =
-      filter_chain.filters_size() == 0 ? "" : filter_chain.filters(0).DebugString();
-  const std::string hcm_str =
-      "type.googleapis.com/"
-      "envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager";
-  if (is_quic && (filter_chain.filters().size() != 1 ||
-                  filter_chain.filters(0).typed_config().type_url() != hcm_str ||
```
Comment (danzh2010) [inline source/server/listener_manager_impl.cc]: Why removing HCM check?

File: source/server/listener_manager_impl.cc [hunk 3]
```diff
@@ -930,6 +930,17 @@ Network::DrainableFilterChainSharedPtr ListenerFilterChainFactoryBuilder::buildF
                                      "{}. \nUse QuicDownstreamTransport instead.",
                                      transport_socket.DebugString()));
   }
+  const std::string config_str =
+      filter_chain.filters_size() == 0 ? "" : filter_chain.filters(0).DebugString();
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc:934]: This isn't used anymore?

**Review by danzh2010 — APPROVED** [review]
LGTM! Thanks for plumbing quic thru network filter config!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**